server.go

```diff
@@ -787,3 +793,12 @@ func (w *response) Close() error {
 	}
 	return nil
 }
+
+// ConnectionState() implements the ConnectionStater.ConnectionState() interface.
+func (w *response) ConnectionState() *tls.ConnectionState {
+	if v, ok := w.tcp.(*tls.Conn); ok {
```

Comment:
I think we should avoid using a concrete type here and, like [x/net/http2](https://github.com/golang/net/blob/922f4815f713f213882e8ef45e0d315b164d705c/http2/server.go#L386), use [an interface](https://github.com/golang/net/blob/922f4815f713f213882e8ef45e0d315b164d705c/http2/http2.go#L332). Perhaps something like:
```go
func (w *response) ConnectionState() *tls.ConnectionState {
	type tlsConnectionStater interface {
		ConnectionState() tls.ConnectionState
	}
	if v, ok := w.tcp.(tlsConnectionStater); ok {
		t := v.ConnectionState()
		return &t
	}
	return nil
} 
```

That way the `*tls.Conn` can be wrapped by a custom listener.

Reply:
ok, I am fine with that. @miekg @tmthrgd is there any other feedback on this diff you want me to push in?